src/App.tsx

```diff
@@ -14,11 +18,100 @@ const goodsFromServer: string[] = [
   'Garlic',
 ];
 
-const App: React.FC = () => (
-  <div className="App">
-    <h1>Goods</h1>
-    {goodsFromServer.length}
-  </div>
-);
+type State = {
+  goods: string[];
+  initialGoods: string[];
+  buttonVisible: boolean;
+};
+
+type Props = {};
+
+class App extends React.Component<Props, State> {
+  state: State = {
+    goods: goodsFromServer,
+    initialGoods: goodsFromServer,
+    buttonVisible: true,
+  };
+
+  start = () => {
+    this.setState((state) => ({
+      goods: state.initialGoods,
+    }));
+  };
+
+  reverse = () => {
+    this.setState((state) => ({
+      goods: [...state.goods].reverse(),
+    }));
+  };
+
+  sort = () => {
+    this.setState((state) => ({
+      goods: [...state.goods].sort(),
+    }));
+  };
+
+  sortByLength = () => {
+    this.setState((state) => ({
+      goods: [...state.goods].sort((a, b) => a.length - b.length),
+    }));
+  };
```

Comment:
As I understood your handlers shouldn't change array, handlers should change some simple fields of state.
Better change copy of array in render method.